File: recyclerview/src/main/java/com/example/administrator/recyclerview/MainActivity.java
package com.example.administrator.recyclerview;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    RecyclerView recyclerView;
    List<Integer> datas;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initDatas();
        recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false));
        MySnapHelper helper = new MySnapHelper(3);
        helper.attachToRecyclerView(recyclerView);
        recyclerView.setAdapter(new RecyclerViewAdapter());
    }

    private void initDatas() {
        datas = new ArrayList<>();
        datas.add(R.mipmap.a);
        datas.add(R.mipmap.b);
        datas.add(R.mipmap.c);
        datas.add(R.mipmap.d);
        datas.add(R.mipmap.e);
        datas.add(R.mipmap.f);
        datas.add(R.mipmap.g);
        datas.add(R.mipmap.h);
        datas.add(R.mipmap.a);
        datas.add(R.mipmap.b);
        datas.add(R.mipmap.c);
        datas.add(R.mipmap.d);
        datas.add(R.mipmap.e);
        datas.add(R.mipmap.f);
        datas.add(R.mipmap.g);
        datas.add(R.mipmap.h);
    }


    class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder> {
        @Override
        public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_image, parent, false));
        }

        @Override
        public void onBindViewHolder(ViewHolder holder, int position) {
            holder.imageView.setImageResource(datas.get(position));
        }

        @Override
        public int getItemCount() {
            return datas.size();
        }

        class ViewHolder extends RecyclerView.ViewHolder {
            ImageView imageView;

            public ViewHolder(View itemView) {
                super(itemView);
                imageView = itemView.findViewById(R.id.image);
            }
        }
    }
}

File: colorfilter/src/main/java/com/example/administrator/colorfilter/MainActivity.kt
package com.example.administrator.colorfilter

import android.animation.ArgbEvaluator
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.graphics.drawable.Drawable
import android.os.Build
import android.os.Bundle
import android.support.annotation.RequiresApi
import android.support.v4.app.ActivityCompat
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.TypedValue
import android.view.View
import android.widget.ImageView
import android.widget.RelativeLayout
import com.example.administrator.colorfilter.divider.DividerBuilder


class MainActivity : AppCompatActivity() {
    var headBg: Drawable? = null
    var iv1Drawable: Drawable? = null
    var iv2Drawable: Drawable? = null

    var rv: RecyclerView? = null
    var rlHead: RelativeLayout? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        /**
         * 初始化工作
         */
        rv = findViewById(R.id.rv)
        rv?.setLayoutManager(LinearLayoutManager(this))
        rv?.setAdapter(SimpleAdapter())

        rlHead = findViewById(R.id.rl_head)

        headBg = rlHead?.getBackground()!!.mutate()//获取head的背景drawable
        iv1Drawable = (findViewById<View>(R.id.iv1) as ImageView).getDrawable().mutate()//获取图片的drawable
        iv2Drawable = (findViewById<View>(R.id.iv2) as ImageView).getDrawable().mutate()//获取图片的drawable


        //给recyclerView一个滚动监听
        rv?.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
                val fraction = calcFraction(dy)

                setUI(fraction)

            }


        })


        val builder = DividerBuilder()
                .divider(ResourcesCompat.getDrawable(resources, R.drawable.shape_divider_normal, null))
                .showBottom(false)
                .showTop(false)
        rv?.addItemDecoration(builder.build())
    }

    private var scrollY: Int = 0

    /**
     *
     * @param dy
     * @return   0~1 ,滑动距离越大，值越大
     */
    private fun calcFraction(dy: Int): Float {
        //这里的300是图片的高度，图片滚完后就是100%
        val imgHeight = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 300F, resources.displayMetrics)
        val toolbarHeight = rlHead?.getHeight()
        val maxHeight = imgHeight - toolbarHeight!!;//图片从头到尾移动的距离

        scrollY += dy //dy是这次移动的距离，每次移动的距离加起来就是总移动的距离，dy是有正有负的

        /**
         *
         */
        return if (scrollY >= maxHeight) {
            1.0f

        } else if (scrollY <= 0) {
            0f
        } else {
            //            float delta = Math.abs(scrollY - maxHeight) ;
            scrollY / maxHeight
        }
    }


    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    private fun setUI(fraction: Float) {
        //背景只需要设置透明度，255是全不透明
        headBg?.setAlpha((fraction * 255).toInt())

        //This evaluator can be used to perform type interpolation between integer values that represent ARGB colors.
        //这个求值器用来执行计算用整形表示的颜色的差值
        val argbEvaluator = ArgbEvaluator()
        val startColor = ActivityCompat.getColor(this@MainActivity, R.color.colorPrimary)
        val endColor = Color.WHITE

        //根据fraction计算出开始和结束中间的色值
        val calcColor = argbEvaluator.evaluate(fraction, startColor, endColor) as Int

        val colorFilter = PorterDuffColorFilter(calcColor, PorterDuff.Mode.SRC_IN)

        iv1Drawable?.setColorFilter(colorFilter)
        iv2Drawable?.setColorFilter(colorFilter)


    }


}
